### PR TITLE
Subtracting curves

### PR DESCRIPTION
**Pull request by matkrak**
As the description says. First part where user doesn't want to create new object
Just 1 question about it:
I was thinking like this: 
If I call the method of object c1 like `c1.subtract(c2, newobj=False)` I don't want it to change the domain of c1. 
The idea of broadcasting says, that the dimension of the output should be max of dimensions on the input. If you want, I can make it like this, and I absolutely will do this in subtract(c1, c2), but here I decided to solve it like I did.
Please give me some feedback. I am working on the second part that is more confusing since when creating new object I need to figure out what the domain should be and I need to care about metadata here as well.

### CODE REVIEW

File: beprof/curve.py [hunk 1]
```diff
@@ -164,23 +189,31 @@ def main():
...
+
+    k2 = k.view(np.ndarray)
+    print(k2)
+
+    k3 = k[1:2,:]
+    print(k3)
+
+    print('\nSubtract method :\n')
+
+    c = Curve([[0.0, 0], [5, 5], [10, 0]])
```
Comment (matkrak) [inline beprof/curve.py:243]: Also here something temporary:
I forced datatype to be float, if I just put 0 instead of 0.0 the result will be wrong, because 
if datatype of self.x is int, operation in line 165 will convert to int as well, and in some cases should be float instead.

**Comment by matkrak** [comment]
About the second part: 
Since the program is here, that means newobj is True
what means the method is not going to modify self but create new curve object instead

The first thing: what should be the domain of a new object?
My proposition for this method is: user can set domain parameter to:
        'keep' - new object domain is self.x
        'extend' - new object domain is sum of sets self.x, c2.x,
                       and values of c2.y in self.x-c2.x and self.x in c2.x - self.x are defval
extend works nice when domains are disjoint. If the domains have a common part what should be done?
The parts of sections that are unique should probably be treated as in 'extend'. What about the common part?
i.e. 
domain1 = [0, 1, 2, 3, 4, 5]
domain2 = [2, 2.5, 3, 3.5, 5, 6, 7, 8]
what should be the new object domain assuming 'extend' option?
I'd say either
[0, 1, 2, 3, 4, 5, 6 ,7 ,8] - the common part is taken from the orginal domain, or
[0, 1, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8] - sum of sets ?

**Comment by matkrak** [comment]
Added next part of subtract method.
This part creates new object, and can handle metadata (however since its object method, it will only respect metadata assigned to this object)
2 more cases left:
Can't really explain in in plain english, but this thing below is even better.
Domains of curves  c1: [a1, b1],     c2:[a2, b2].
For now I solved the case
------a1||||||b1-----a2|||||b2------>            this is X axis (and also I solved mirrored case when a2b2 comes first)
Todo:
------a1|||||a2||b1||||||b2---------->                            so there is common part
------a1||||||b1|||||||||a2|||||b2---->                    or one domain includes the other.

The thing I mentioned before (float to int conversion) is still unsolved but I think I will deal with it after finishing this part. Please give me some feedback

File: beprof/curve.py [hunk 2]
```diff
@@ -105,7 +106,7 @@ def change_domain(self, domain):
             print('Error2')
             return self
         y = np.interp(domain, self.x, self.y)
-        obj = Curve(np.stack((domain, y), axis=1))
+        obj = Curve(np.stack((domain, y), axis=1), **self.__dict__['metadata'])
```
Comment (grzanka) [inline beprof/curve.py]: Why this stuff is here ? It is not related to implementation of #26

File: beprof/curve.py [hunk 3]
```diff
@@ -143,6 +144,68 @@ def rebinned(self, step=0.1, fixp=0):
         domain = [fixp + n * step for n in range(int(count_start), int(count_stop)+1)]
         return self.change_domain(domain)
 
+    def subtract(self, curve2, defval=0, newobj=True, domain='keep'):
```
Comment (grzanka) [inline beprof/curve.py]: Add docstring explaining meaning of parameters

File: beprof/curve.py [hunk 4]
```diff
@@ -143,6 +144,68 @@ def rebinned(self, step=0.1, fixp=0):
...
+                if a2 > b1:     # here we have ------a1||||||b1-----a2|||||b2------>
+                    newX = np.concatenate((self.x, curve2.x))        # new X is just an union
+                    newY = np.concatenate((tmp1, tmp2))    # into one list
+                else:           # mirrored, just a minor change
+                    newX = np.concatenate((curv2.x, self.x))        # new X is just an union
+                    newY = np.concatenate((tmp2, tmp1))             # into one list
+
+                print('newX: ', newX)
+                print('newY: ', newY)
+                obj = Curve(np.stack((newX, newY), axis=1), **self.__dict__['metadata'])
```
Comment (grzanka) [inline beprof/curve.py]: Will this work if I try to subtract two profiles ?

```
p1 = profile.Profile(...)
p2 = profile.Profile(...)
r = p1.subtract(p2)
```

What will be the type of r ?

File: beprof/curve.py (hunk 2 above)
Comment (matkrak) [inline beprof/curve.py]: wanted to be sure substraction can handle metadata. Will be removed

File: beprof/curve.py (hunk 4 above)
Comment (matkrak) [inline beprof/curve.py]: didn't try profiles yet. I want to finish the function for curves for now. 
But if you ask, p1.subtract(p2) will return an object class type(p1) in the future

**Comment by matkrak** [comment]
Just realised that this function as it is doesn't make any sense. I am going to rewrite it, and what I mean more precisely is behaviors depending on the 2 parameters.
1) newobj should just decide if the method should modify self or create new obj.
2) domain should decide what should be domain of the returned value.

All 4 combinations (assuming 2 values of domain) can be done without throwing exception. 
Also every time there are 3 cases I described above

File: beprof/curve.py (hunk 4 above)
Comment (grzanka) [inline beprof/curve.py]: OK.

**Comment by grzanka** [comment]
What do you think about such thing:
1.  add a method to class Curve which return callable object which can give you back interpolated values for single argument or for a list of arguments (np.ndarray) ? It can be called evaluate_at_x for sth more reasonable.
2. numpy has two method for subtracting: np.subtract (works like "-") and oveloaded `__sub__` (which works like "-=").  We could follow similar approach and have beprof.subtract method which always returns new object and Curve.subtract (working like "-=") which sometimes returns new object and sometimes not.

In this way you could more easily separate cases:
- beprof.subtract would deal with all the strange cases of disjoint domains etc...
- Curve.subtract would work only in the case when domain of self is in domain of the argument. By default Curve.subtract returns None. If user specifies some flag it returns new object with result of subtraction (you can use here beprof.subtract) and self stays untouched. 

Is it reasonable ?

**Comment by matkrak** [comment]
Yes,, perfectly reasonable. I have been wondering If I should add a method like evaluate_at_x you mentioned, because that was was a piece of code that would appear in several places in the method.
I will try to figure it out now.

**Comment by matkrak** [comment]
So I started writing subtract fucntion in functions.py and I realised that it doesn't really matter what are these domains of curve1 and curve2, because I can always solve it like I did. I commented more efficient solution I think, but this one that left is just universal. It uses sets and sort() method what can make it slow if the dataset is really big. Should I use them or come up with more efficient solution as I started (I left it commented so that you can see what I mean).
I will be working on logging package since you give me some feedback of what I should do.

File: beprof/functions.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    #     print('newX: ', newX)
+    #     print('newY: ', newY)
+    #     obj = curve.Curve(np.stack((newX, newY), axis=1))  # ADD METADATA HANDLING!
+    #     return obj
+    # # CASE 2: one domain includes the other, in this example domain(curve2) is inside domain(curve1)
+    # elif a1 <= a2 and b1 >= b2:
+    set1 = set(curve1.x)
+    set2 = set(curve2.x)
+    newX = list(set1 | set2)
+    newX.sort()
```
Comment (grzanka) [inline beprof/functions.py]: I know its compact, but inefficient in term of memory (creating two new sets and one new list) and also in term of computing (sorting union of two sorted lists).
Maybe reading about merge-sort will give you some inspiration ?

I am not sure if this will be helpful, but consider also numpy.searchsorted. Investigate also complexity of this algorithm

File: beprof/functions.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    #
+    #     print('newX: ', newX)
+    #     print('newY: ', newY)
+    #     obj = curve.Curve(np.stack((newX, newY), axis=1))  # ADD METADATA HANDLING!
+    #     return obj
+    # # CASE 2: one domain includes the other, in this example domain(curve2) is inside domain(curve1)
+    # elif a1 <= a2 and b1 >= b2:
+    set1 = set(curve1.x)
+    set2 = set(curve2.x)
+    newX = list(set1 | set2)
```
Comment (grzanka) [inline beprof/functions.py]: And what about this http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.union1d.html ?

Comment (grzanka) [inline beprof/functions.py]: Check if union1d gives sorted result

Comment (matkrak) [inline beprof/functions.py]: union1d gives sorted results and since separatly the domains are sorted, it is probably going to merge it in O(n) time. Way better solution.

**Comment by matkrak** [comment]
Can I leave this algoritm as it is now, and focus on futher tasks in this issue?

File: beprof/functions.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    #         newY = np.concatenate((tmp2, tmp1))             # into one list
+    #
+    #     print('newX: ', newX)
+    #     print('newY: ', newY)
+    #     obj = curve.Curve(np.stack((newX, newY), axis=1))  # ADD METADATA HANDLING!
+    #     return obj
+    # # CASE 2: one domain includes the other, in this example domain(curve2) is inside domain(curve1)
+    # elif a1 <= a2 and b1 >= b2:
+
+    newX = np.union1d(curve1.x, curve2.x)
```
Comment (grzanka) [inline beprof/functions.py]: Read PEP 0008 to learn whats wrong in naming variables like that

File: beprof/functions.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    print('newX: ', newX)
+    y1 = curve1.evaluate_at_x(newX, defval)
+    y2 = curve2.evaluate_at_x(newX, defval)
+    newY = y1 - y2
+    obj = curve.Curve(np.stack((newX, newY), axis=1))  # ADD METADATA HANDLING!
```
Comment (grzanka) [inline beprof/functions.py]: What if we subtract two profiles ? Resulting type will be OK ?

Comment (matkrak) [inline beprof/functions.py]: for now - no. I need to find out how to call **new** method of object depending on curve1 and curve2

**Comment by grzanka** [comment]
By the way - could you merge actual state of master branch into your working branch for this PR (or better do the rebase ) ?

File: beprof/functions.py (hunk 8 above)
Comment (matkrak) [inline beprof/functions.py]: Is this a good solution if I wanted providing I want obj to be of the same class as curve1?
`obj = curve1.__class__(np.stack((newX, newY), axis=1))`

And more important question: which class should be obj? There could be a parameter by default set to 'same_as_curve1' class, or somethink like that, and one could change it to 'same_as_curve_2', or maybe some other solution?
What about this docstring I made to subtract method of Curve class? Is the idea OK?

**Comment by matkrak** [comment]
Sure I just need to find a way to do this, I am a bit confused about this merging stuff

**Comment by matkrak** [comment]
Almost succeeded, didn't remove conflicts, doing it right now

File: beprof/curve.py [hunk 9]
```diff
@@ -144,6 +146,23 @@ def rebinned(self, step=0.1, fixp=0):
         domain = [fixp + n * step for n in range(int(count_start), int(count_stop)+1)]
         return self.change_domain(domain)
 
+    def evaluate_at_x(self, arg, defval=0):
```
Comment (grzanka) [inline beprof/curve.py]: Add some docstring here

File: beprof/curve.py [hunk 10]
```diff
@@ -55,6 +56,7 @@ def __array_finalize__(self, obj):
             return
         self.metadata = getattr(obj, 'metadata', {})
 
+
```
Comment (grzanka) [inline beprof/curve.py]: Clean unnecesary empty lines

File: beprof/functions.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    #     else:           # mirrored, just a minor change
+    #         newX = np.concatenate((curve2.x, curve1.x))        # new X is just an union
+    #         newY = np.concatenate((tmp2, tmp1))             # into one list
+    #
+    #     print('newX: ', newX)
+    #     print('newY: ', newY)
+    #     obj = curve.Curve(np.stack((newX, newY), axis=1))  # ADD METADATA HANDLING!
+    #     return obj
+    # # CASE 2: one domain includes the other, in this example domain(curve2) is inside domain(curve1)
+    # elif a1 <= a2 and b1 >= b2:
```
Comment (grzanka) [inline beprof/functions.py]: This alternative implementation could go to separate branch (might be also branch from branch) in your fork of my repo. If it stays there, it can be later recuperated (merged into) if needed.

File: beprof/functions.py (hunk 8 above)
Comment (grzanka) [inline beprof/functions.py]: If you write it clearly in docstring, the idea to inherit type from curve1 seems OK.
Have you checked what happens if you call np.subtract on two objects of Curve and Profile classes ?
I would stay this time without a parameter and create an issue to add it later.

Docstring seems OK.

**Comment by matkrak** [comment]
I made changes you marked, now I am going to write subtract method of Curve class as described in the docstring there.

**Comment by matkrak** [comment]
Wrote subtract method in curve function. Also made a small change mentioned in the last commit.
Is it OK or it needs changes? Please let me know.
The final version needs docstring tests instead of these prints in main() I do realise that. Just waiting for confirmation that everything is as it should be for now.

File: beprof/curve.py [hunk 12]
```diff
@@ -144,6 +147,52 @@ def rebinned(self, step=0.1, fixp=0):
...
+            print('Error - curve2 domain does not include self domain')
+            return None
+        # if one want to create and return a new object rather then modify self
+        if newobj:
+            return functions.subtract(self, curve2.change_domain(self.x))
+        # important comment about the line above. Need to modify change_domain to be more
+        # universal - for now it can only return a Curve. Same trick as in functions.subtract
+        # is needed here so that change_domain can obj type of any subclass of curve as well
+        values = curve2.evaluate_at_x(self.x)
+        self.y = self.y - values
```
Comment (grzanka) [inline beprof/curve.py:227]: What about using `self.y -= values` ?

Comment (matkrak) [inline beprof/curve.py:227]: Will be fine as well

**Comment by matkrak** [comment]
Can I remove unnecessary comments and write docstring tests for both subtract function and method to close this PR, or there is something that needs to be fixed? ?

**Comment by grzanka** [comment]
I will check it tomorrow You can now remove comments and write docstrings.
Remember, some other tasks are also pending and might also be time consuming.

File: beprof/curve.py [hunk 13]
```diff
@@ -144,6 +147,70 @@ def rebinned(self, step=0.1, fixp=0):
...
+        [ 100.    1.    2.    3.  100.]
+
+        :param arg: x-value to calculate Y (may be an array or list as well)
+        :param defval: default value to return if can't interpolate value at arg
+        :return: np.array of Y-values at arg. If arg is a scalar, will return scalar as well
+        '''
+        y = np.interp(arg, self.x, self.y, left=defval, right=defval)
+        return y
+
+    def subtract(self, curve2, newobj=False,):
```
Comment (grzanka) [inline beprof/curve.py]: What does this comma after False ?

Comment (matkrak) [inline beprof/curve.py]: Nothing I didn't notice it. I wasn't even aware the program will run with such a thing.

**Comment by matkrak** [comment]
Ready to merge?

File: beprof/curve.py [hunk 14]
```diff
@@ -107,6 +108,8 @@ def change_domain(self, domain):
             return self
         y = np.interp(domain, self.x, self.y)
         obj = Curve(np.stack((domain, y), axis=1), **self.__dict__['metadata'])
+        # as mentioned below, should be as in the line below instead
+        # obj = self.__class__(np.stack((domain, y), axis=1), **self.__dict__['metadata'])
```
Comment (grzanka) [inline beprof/curve.py]: So replace it with the line below

**Comment by grzanka** [comment]
It looks you need to fix importing structure in main functions.
Probably in PyCharm it works, but not in command line.

You should be also able to run from command line `python3 curve.py` (or equvalent command) and obtain reasonable results.

Read more about relative import in Python - its pretty complex topic (it was also discussed during lectures)

```
>>> import beprof.curve as c
>>> c.main()

Subtract method :

c: 
 shape: (3, 2)
X : [0.000,10.000]
Y : [0.000000,5.000000]
Metadata : {}
d: 
 shape: (3, 2)
X : [0.000,10.000]
Y : [1.000000,1.000000]
Metadata : {}
c-d: 
 shape: (3, 2)
X : [0.000,10.000]
Y : [-1.000000,4.000000]
Metadata : {}
*********
X: [  0.   5.  10.]
Y: [-1.  4. -1.]
Newobj creation #1



a: 

X:  [0 1 2 3]
Y:  [0 1 2 1]
M:  {'positive': 'example'}

b: 

X:  [ 0.5  1.5  2.   2.5]
Y:  [ 1.  1.  1.  1.]
M:  {'negative': 'one'}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/grzanka/PycharmProjects/beprof/beprof/curve.py", line 251, in main
    diff = functions.subtract(a, b)
NameError: name 'functions' is not defined
```

**Comment by matkrak** [comment]
I am using Pycharm, where instead of `import beprof.functions` I have `import functions`.
File is in the same directory and works with no errors.
Also when I try python from terminal opened in the right directory it works fine.
Should I change it to `import functions` or I should read more and then try to fix it ?

**Comment by grzanka** [comment]
Make it in such way that it works in following situations:
- beprof installed with pip
- command line execution of curve.py script
- running script curve.py from PyCharm

**Comment by grzanka** [comment]
http://docs.python-guide.org/en/latest/writing/structure/
might be useful to read

File: beprof/profile.py [hunk 15]
```diff
@@ -1,5 +1,6 @@
 from beprof.curve import Curve
 import numpy as np
+from beprof import functions as f
```
Comment (grzanka) [inline beprof/profile.py]: I see some contraction in importing Curve and functions.
First one you do using `from x.y import z` and the second one you do using `from x import y as z`. Why ?

File: beprof/profile.py [hunk 16]
```diff
@@ -195,37 +196,28 @@ def plateau_to_max(self):
 
 def main():
     print('\nProfile')
-    p = Profile([[1, 0], [2, 1], [3, 0]], first='one', second='two', third='three')
+    p = Profile([[0, 0], [1, 1], [2, 2], [3, 1]], some='exapmlary', meta='data')
```
Comment (grzanka) [inline beprof/profile.py]: exapmlary?

File: beprof/curve.py [hunk 17]
```diff
@@ -144,6 +145,70 @@ def rebinned(self, step=0.1, fixp=0):
...
+            return None
+        # if one want to create and return a new object rather then modify self
+        if newobj:
+            return f.subtract(self, curve2.change_domain(self.x))
+        values = curve2.evaluate_at_x(self.x)
+        self.y = self.y - values
+        return None
+
+
+
```
Comment (grzanka) [inline beprof/curve.py]: How many spaces should be used to spearate methods ?

**Comment by grzanka** [comment]
Still doesn't work (beprof not installed using pip):

```
C:\Users\User\PycharmProjects\beprof\beprof>python curve.py
Traceback (most recent call last):
  File "curve.py", line 8, in <module>
    from beprof import functions as f
ImportError: No module named 'beprof'

C:\Users\User\PycharmProjects\beprof\beprof>cd ..

C:\Users\User\PycharmProjects\beprof>python beprof\curve.py
Traceback (most recent call last):
  File "beprof\curve.py", line 8, in <module>
    from beprof import functions as f
ImportError: No module named 'beprof'

C:\Users\User\PycharmProjects\beprof>python
Python 3.4.3 |Anaconda 2.3.0 (64-bit)| (default, Mar  6 2015, 12:06:10) [MSC v.1600 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from beprof import curve
>>> c = curve.Curve([[1,0]])
>>> c
Curve([[1, 0]])

C:\Users>python
Python 3.4.3 |Anaconda 2.3.0 (64-bit)| (default, Mar  6 2015, 12:06:10) [MSC v.1600 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from beprof import curve
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named 'beprof'
>>>
```

File: beprof/curve.py [hunk 18]
```diff
@@ -5,6 +5,7 @@
 from scipy import signal
 import math
 import copy
+from beprof import functions as f
```
Comment (grzanka) [inline beprof/curve.py]: Importing style `from x import y as z` was designed to deal with two cases: name clash or very long module names. It is not our case. Read:
- https://code.google.com/p/soc/wiki/PythonStyleGuide#Imports
- http://docs.python-guide.org/en/latest/writing/structure/#modules

**Comment by grzanka** [comment]
Fix conflicts first.

**Comment by matkrak** [comment]
Solved conflicts, ready to merge finally

**Comment by grzanka** [comment]
Provide a command which I can use to run main function from curve.py file to test if everything implemented there works correctly.

First, read this: http://ncoghlan-devs-python-notes.readthedocs.org/en/latest/python_concepts/import_traps.html

**Comment by matkrak** [comment]
I skimmed the article you mentioned. I am using ipython to test modules:

```
ipython
import beprof.curve as c
c.main()
```

How would you like it to be done?
Also: do I need separate **init.py** file inside the main directory?

**Comment by grzanka** [comment]
First of all - read this article carefully, not just skim it.

Try command line, instead of IPython.
In my case it doesn't work:

```
test@ce3b6ada1d29:~/beprof$ ls
LICENSE  README.md  beprof  setup.py
test@ce3b6ada1d29:~/beprof$ python3
Python 3.4.3 (default, Oct 14 2015, 20:28:29) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import beprof.curve as c
>>> c.main()

Subtract method :

c: 
 shape: (3, 2)
X : [0.000,10.000]
Y : [0.000000,5.000000]
Metadata : {}
d: 
 shape: (3, 2)
X : [0.000,10.000]
Y : [1.000000,1.000000]
Metadata : {}
c-d: 
 shape: (3, 2)
X : [0.000,10.000]
Y : [-1.000000,4.000000]
Metadata : {}
*********
X: [  0.   5.  10.]
Y: [-1.  4. -1.]
Newobj creation #1



a: 

X:  [0 1 2 3]
Y:  [0 1 2 1]
M:  {'positive': 'example'}

b: 

X:  [ 0.5  1.5  2.   2.5]
Y:  [ 1.  1.  1.  1.]
M:  {'negative': 'one'}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/test/beprof/beprof/curve.py", line 268, in main
    diff = functions.subtract(a, b)
  File "/home/test/beprof/beprof/functions.py", line 18, in subtract
    obj = curve1.__class__(np.stack((coord1, coord2), axis=1), **curve1.__dict__['metadata'])
AttributeError: 'module' object has no attribute 'stack'
```

It should be possible to run main method from one single command which I copy-paste into my bash shell. 

Hint: read about "python -m".

**Comment by matkrak** [comment]
I must read more about it, what makes it harder to understand is it works for me perfectly. Both the way I described before and 

```
python3 -m beprof.curve
python3 -m beprof.profile
```

works just fine (main is executed in both cases, I noticed one line in profile.py is missing - Profile constructor doesn't return anything, I will restore it in the next commit).

**Comment by grzanka** [comment]
My fault, I had old numpy version installed.

```
test@5c4160e4b8b4:~/beprof$ python3
Python 3.4.3 (default, Oct 14 2015, 20:28:29) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
>>> numpy.version.version
'1.8.2'
```

With numpy 1.10 it works OK.
